File: ruby/loadbalancer/RoundRobinBalancer.rb
require 'socket'
require 'timeout'
require 'yaml'
require_relative 'server'

class RoundRobinBalancer
  def initialize(servers, timeout: 5)
    @servers = servers.map { |server| parse_backend_address(server) }
    @server_enum = @servers.cycle
    @timeout = timeout
  end

  def next_server
    raise "No healthy servers available" if @servers.none?(&:healthy)

    loop do
      server = @server_enum.next
      return server if server.healthy
    end
  end

  def health_check
    @servers.each do |backend_server|
      backend_server.check_health(@timeout)
    end
  end

  private

  def parse_backend_address(server)
    ip, port = server.split(':')
    Server.new(ip, port)
  end
end

File: go/loadbalancer/round_robin_balancer.go
package main

import (
	"errors"
	"strings"
	"sync/atomic"
	"time"
)

type RoundRobinBalancer struct {
	servers []*Server
	index   uint32
	timeout time.Duration
}

func NewRoundRobinBalancer(servers []string, timeout time.Duration) *RoundRobinBalancer {
	backendServers := make([]*Server, len(servers))
	for i, server := range servers {
		backendServers[i] = parseBackendAddress(server)
	}
	return &RoundRobinBalancer{servers: backendServers, timeout: timeout}
}

func (rrb *RoundRobinBalancer) NextServer() (*Server, error) {
	for i := 0; i < len(rrb.servers); i++ {
		index := atomic.AddUint32(&rrb.index, 1) - 1
		server := rrb.servers[index%uint32(len(rrb.servers))]
		if server.Healthy {
			return server, nil
		}
	}
	return nil, errors.New("no healthy servers available")
}

func (rrb *RoundRobinBalancer) HealthCheck() {
	for _, server := range rrb.servers {
		server.CheckHealth(rrb.timeout)
	}
}

func (rrb *RoundRobinBalancer) HealthyServerCount() int {
	count := 0
	for _, server := range rrb.servers {
		if server.Healthy {
			count++
		}
	}
	return count
}

func parseBackendAddress(server string) *Server {
	parts := strings.Split(server, ":")
	return NewServer(parts[0], parts[1])
}
